src/transform-sdk/cpp/src/transform_sdk.cc

```diff
@@ -0,0 +1,632 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <redpanda/transform_sdk.h>
+
+#include <algorithm>
+#include <chrono>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <expected>
+#include <system_error>
+#include <utility>
+
+namespace redpanda {
+namespace {
+
+// NOLINTBEGIN(*-macro-usage,*-macro-parentheses)
+
+#define RP_CONCAT(x, y) RP_CONCAT_IMPL(x, y)
+#define RP_CONCAT_IMPL(x, y) x##y
+
+#define ASSIGN_OR_RETURN(lhs, rexpr)                                           \
+    RP_ASSIGN_OR_RETURN_IMPL(RP_CONCAT(expected_tmp_, __LINE__), lhs, rexpr);
+
+#define RP_ASSIGN_OR_RETURN_IMPL(uniq_var_name, lhs, rexpr)                    \
+    auto uniq_var_name = (rexpr);                                              \
+    if (!uniq_var_name.has_value()) [[unlikely]] {                             \
+        return std::unexpected(uniq_var_name.error());                         \
+    }                                                                          \
+    lhs = std::move(*uniq_var_name)
```

Comment:
kinda feels like this could be done monadically and avoid the macro, but idk

Reply:
It could be, but it's personally more noise IMO.